File: macos/PlaylistFolderDownloader/Sources/PlaylistFolderDownloader/BackendClient.swift
import Foundation

final class ProcessRegistry: @unchecked Sendable {
    private let lock = NSLock()
    private var processes: [UUID: Process] = [:]

    func insert(_ process: Process) -> UUID {
        let id = UUID()
        lock.lock()
        processes[id] = process
        lock.unlock()
        return id
    }

    func remove(_ id: UUID) {
        lock.lock()
        processes[id] = nil
        lock.unlock()
    }

    func terminateAll() {
        lock.lock()
        let runningProcesses = Array(processes.values)
        lock.unlock()

        for process in runningProcesses where process.isRunning {
            process.terminate()
        }
    }
}

enum BackendError: LocalizedError {
    case repoRootNotFound
    case invalidOutput
    case processFailed(String)

    var errorDescription: String? {
        switch self {
        case .repoRootNotFound:
            "Could not find the project root containing pyproject.toml."
        case .invalidOutput:
            "The Python backend returned invalid output."
        case let .processFailed(message):
            message
        }
    }
}

actor BackendClient {
    private let decoder = JSONDecoder()
    private let encoder = JSONEncoder()
    private let processRegistry = ProcessRegistry()

    func load(url: String) async throws -> PlaylistInfo {
        let output = try await runBackend(arguments: ["load", url])
        guard let line = output.split(separator: "\n").last else {
            throw BackendError.invalidOutput
        }
        let data = Data(line.utf8)
        if let failure = try? decoder.decode(FailureEnvelope.self, from: data),
           failure.event == "failed" {
            throw BackendError.processFailed(failure.error ?? "Could not load metadata.")
        }
        return try decoder.decode(LoadEnvelope.self, from: data).playlist
    }

    func probe(video: VideoInfo) async throws -> VideoInfo {
        let videoData = try encoder.encode(video)
        let videoJSON = String(decoding: videoData, as: UTF8.self)
        let output = try await runBackend(arguments: ["probe", videoJSON])
        guard let line = output.split(separator: "\n").last else {
            throw BackendError.invalidOutput
        }
        let data = Data(line.utf8)
        if let failure = try? decoder.decode(FailureEnvelope.self, from: data),
           failure.event == "failed" {
            throw BackendError.processFailed(failure.error ?? "Could not probe video formats.")
        }
        return try decoder.decode(ProbeEnvelope.self, from: data).video
    }

    func download(
        request: DownloadRequest,
        onEvent: @escaping @MainActor (DownloadEvent) -> Void
    ) async throws {
        let requestData = try encoder.encode(request)
        let process = try makeProcess(arguments: ["download"])
        let processID = processRegistry.insert(process)
        defer { processRegistry.remove(processID) }

        let stdin = Pipe()
        let stdout = Pipe()
        let errPipe = Pipe()
        process.standardInput = stdin
        process.standardOutput = stdout
        process.standardError = errPipe

        try process.run()
        stdin.fileHandleForWriting.write(requestData)
        stdin.fileHandleForWriting.closeFile()

        Task.detached {
            let data = errPipe.fileHandleForReading.readDataToEndOfFile()
            if let text = String(data: data, encoding: .utf8), !text.isEmpty {
                fputs(text, Darwin.stderr)
            }
        }

        for try await line in stdout.fileHandleForReading.bytes.lines {
            if Task.isCancelled {
                processRegistry.terminateAll()
                throw CancellationError()
            }
            guard !line.isEmpty else { continue }
            if let data = line.data(using: .utf8),
               let event = try? decoder.decode(DownloadEvent.self, from: data) {
                await onEvent(event)
            }
        }

        process.waitUntilExit()
        if process.terminationStatus != 0 {
            if Task.isCancelled || process.terminationReason == .uncaughtSignal {
                throw CancellationError()
            }
            throw BackendError.processFailed("The Python backend exited with status \(process.terminationStatus).")
        }
    }

    nonisolated func cancelDownload() {
        processRegistry.terminateAll()
    }

    nonisolated func cancelAllOperations() {
        processRegistry.terminateAll()
    }

    private func runBackend(arguments: [String]) async throws -> String {
        let process = try makeProcess(arguments: arguments)
        let processID = processRegistry.insert(process)
        defer { processRegistry.remove(processID) }

        let stdout = Pipe()
        let errPipe = Pipe()
        process.standardOutput = stdout
        process.standardError = errPipe
        try process.run()

        let output = stdout.fileHandleForReading.readDataToEndOfFile()
        let errorOutput = errPipe.fileHandleForReading.readDataToEndOfFile()
        process.waitUntilExit()

        if let text = String(data: errorOutput, encoding: .utf8), !text.isEmpty {
            fputs(text, Darwin.stderr)
        }
        if process.terminationStatus != 0 {
            if let message = decodeFailureMessage(from: output) {
                throw BackendError.processFailed(message)
            }
            let message = String(data: errorOutput, encoding: .utf8) ?? "Backend failed."
            throw BackendError.processFailed(message)
        }
        return String(data: output, encoding: .utf8) ?? ""
    }

    private func decodeFailureMessage(from data: Data) -> String? {
        guard let output = String(data: data, encoding: .utf8),
              let line = output.split(separator: "\n").last else {
            return nil
        }

        guard let failure = try? decoder.decode(FailureEnvelope.self, from: Data(line.utf8)),
              failure.event == "failed" else {
            return nil
        }

        return failure.error
    }

    private func makeProcess(arguments: [String]) throws -> Process {
        let root = try findRepoRoot()
        let uv = ProcessInfo.processInfo.environment["UV"] ?? "uv"
        let quotedArgs = arguments.map(shellQuote).joined(separator: " ")
        let command = "cd \(shellQuote(root.path)) && \(uv) run python -m playlist_folder_downloader.cli \(quotedArgs)"

        let process = Process()
        process.executableURL = URL(fileURLWithPath: "/bin/zsh")
        process.arguments = ["-lc", command]
        process.environment = ProcessInfo.processInfo.environment.merging(["PYTHONUNBUFFERED": "1"]) { _, new in new }
        return process
    }

    private func findRepoRoot() throws -> URL {
        if let override = ProcessInfo.processInfo.environment["PFD_BACKEND_ROOT"] {
            return URL(fileURLWithPath: override)
        }

        var url = URL(fileURLWithPath: FileManager.default.currentDirectoryPath)
        for _ in 0..<8 {
            if FileManager.default.fileExists(atPath: url.appendingPathComponent("pyproject.toml").path) {
                return url
            }
            url.deleteLastPathComponent()
        }
        throw BackendError.repoRootNotFound
    }

    private func shellQuote(_ value: String) -> String {
        "'" + value.replacingOccurrences(of: "'", with: "'\\''") + "'"
    }
}

File: windows/PlaylistFolderDownloader/BackendClient.cs
using System.Collections.Concurrent;
using System.Diagnostics;
using System.Text.Json;

namespace PlaylistFolderDownloader;

public sealed class BackendClient
{
    private static readonly JsonSerializerOptions JsonOptions = new(JsonSerializerDefaults.Web)
    {
        PropertyNameCaseInsensitive = true,
    };

    private readonly ConcurrentDictionary<Guid, Process> _processes = [];

    public async Task<PlaylistInfo> LoadAsync(string url, CancellationToken cancellationToken)
    {
        FrontendLog.Write($"Load requested: {url}");
        var output = await RunBackendAsync(["load", url], null, cancellationToken);
        var line = LastJsonLine(output) ?? throw new BackendException("The Python backend returned no output.");
        ThrowIfFailure(line, "Could not load metadata.");
        var envelope = JsonSerializer.Deserialize<PlaylistEnvelope>(line, JsonOptions)
            ?? throw new BackendException("The Python backend returned invalid load data.");
        return envelope.Playlist;
    }

    public async Task<VideoInfo> ProbeAsync(VideoInfo video, CancellationToken cancellationToken)
    {
        var videoJson = JsonSerializer.Serialize(new Dictionary<string, object?>
        {
            ["id"] = video.Id,
            ["title"] = video.Title,
            ["webpage_url"] = video.WebpageUrl,
            ["playlist_index"] = video.PlaylistIndex,
        }, JsonOptions);
        var output = await RunBackendAsync(["probe"], videoJson, cancellationToken);
        var line = LastJsonLine(output) ?? throw new BackendException("The Python backend returned no output.");
        ThrowIfFailure(line, "Could not probe video formats.");
        var envelope = JsonSerializer.Deserialize<ProbeEnvelope>(line, JsonOptions)
            ?? throw new BackendException("The Python backend returned invalid probe data.");
        return envelope.Video;
    }

    public async Task DownloadAsync(
        DownloadRequest request,
        Func<DownloadEvent, Task> onEvent,
        CancellationToken cancellationToken)
    {
        var requestJson = JsonSerializer.Serialize(request, JsonOptions);
        using var process = MakeBackendProcess(["download"]);
        var processId = Track(process);
        Task<string>? errorTask = null;
        try
        {
            StartBackendProcess(process);
            await process.StandardInput.WriteAsync(requestJson.AsMemory(), cancellationToken);
            process.StandardInput.Close();

            errorTask = process.StandardError.ReadToEndAsync();

            while (!process.StandardOutput.EndOfStream)
            {
                cancellationToken.ThrowIfCancellationRequested();
                var line = await process.StandardOutput.ReadLineAsync(cancellationToken);
                if (string.IsNullOrWhiteSpace(line))
                {
                    continue;
                }

                var downloadEvent = JsonSerializer.Deserialize<DownloadEvent>(line, JsonOptions);
                if (downloadEvent is not null)
                {
                    await onEvent(downloadEvent);
                }
            }

            await process.WaitForExitAsync(cancellationToken);
            var error = await errorTask;
            if (!string.IsNullOrWhiteSpace(error))
            {
                FrontendLog.Write($"Backend stderr: {error.Trim()}");
                Console.Error.WriteLine(error);
            }

            if (process.ExitCode != 0)
            {
                throw new BackendException(
                    string.IsNullOrWhiteSpace(error)
                        ? $"The Python backend exited with status {process.ExitCode}."
                        : error.Trim());
            }
        }
        catch (OperationCanceledException)
        {
            Terminate(process);
            throw;
        }
        finally
        {
            _processes.TryRemove(processId, out _);
        }
    }

    public void CancelAllOperations()
    {
        foreach (var process in _processes.Values)
        {
            Terminate(process);
        }
    }

    private async Task<string> RunBackendAsync(
        IReadOnlyList<string> arguments,
        string? standardInput,
        CancellationToken cancellationToken)
    {
        using var process = MakeBackendProcess(arguments);
        var processId = Track(process);
        try
        {
            StartBackendProcess(process);
            if (standardInput is not null)
            {
                await process.StandardInput.WriteAsync(standardInput.AsMemory(), cancellationToken);
            }

            process.StandardInput.Close();
            var outputTask = process.StandardOutput.ReadToEndAsync(cancellationToken);
            var errorTask = process.StandardError.ReadToEndAsync(cancellationToken);
            await process.WaitForExitAsync(cancellationToken);
            var output = await outputTask;
            var error = await errorTask;
            if (!string.IsNullOrWhiteSpace(error))
            {
                FrontendLog.Write($"Backend stderr: {error.Trim()}");
                Console.Error.WriteLine(error);
            }

            if (process.ExitCode != 0)
            {
                FrontendLog.Write($"Backend exited with status {process.ExitCode}. Stdout: {output.Trim()}");
                if (LastJsonLine(output) is { } line)
                {
                    ThrowIfFailure(line, "The Python backend failed.");
                }

                throw new BackendException(
                    string.IsNullOrWhiteSpace(error) ? $"The Python backend exited with status {process.ExitCode}." : error.Trim());
            }

            FrontendLog.Write($"Backend completed successfully. Stdout length: {output.Length}");
            return output;
        }
        catch (OperationCanceledException)
        {
            Terminate(process);
            throw;
        }
        finally
        {
            _processes.TryRemove(processId, out _);
        }
    }

    private Process MakeBackendProcess(IReadOnlyList<string> backendArguments)
    {
        var root = FindRepoRoot();
        var python = FindRepoPython(root);
        ProcessStartInfo startInfo;

        if (python is not null)
        {
            startInfo = NewBackendStartInfo(python, root);
        }
        else
        {
            startInfo = NewBackendStartInfo(FindUvExecutable(), root);
            startInfo.ArgumentList.Add("run");
            startInfo.ArgumentList.Add("python");
        }

        startInfo.Environment["PYTHONUNBUFFERED"] = "1";
        startInfo.ArgumentList.Add("-m");
        startInfo.ArgumentList.Add("playlist_folder_downloader.cli");
        foreach (var argument in backendArguments)
        {
            startInfo.ArgumentList.Add(argument);
        }

        return new Process { StartInfo = startInfo, EnableRaisingEvents = true };
    }

    private static ProcessStartInfo NewBackendStartInfo(string executable, DirectoryInfo root)
    {
        return new ProcessStartInfo
        {
            FileName = executable,
            WorkingDirectory = root.FullName,
            RedirectStandardInput = true,
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            UseShellExecute = false,
            CreateNoWindow = true,
        };
    }

    private static void StartBackendProcess(Process process)
    {
        try
        {
            FrontendLog.Write(
                "Starting backend: "
                + $"{process.StartInfo.FileName} {RedactedArguments(process.StartInfo.ArgumentList)} "
                + $"(cwd={process.StartInfo.WorkingDirectory})");
            process.Start();
        }
        catch (Exception ex)
        {
            var fileName = process.StartInfo.FileName;
            var workingDirectory = process.StartInfo.WorkingDirectory;
            throw new BackendException(
                $"Could not start the Python backend. Tried '{fileName}' in '{workingDirectory}'. "
                + $"Make sure uv is installed and run 'uv sync --extra dev'. Details: {ex.Message}");
        }
    }

    private static string RedactedArguments(IEnumerable<string> arguments)
    {
        return string.Join(
            " ",
            arguments.Select(argument => argument.TrimStart().StartsWith('{') ? "<json>" : argument));
    }

    private static DirectoryInfo FindRepoRoot()
    {
        var overrideRoot = Environment.GetEnvironmentVariable("PFD_BACKEND_ROOT");
        if (!string.IsNullOrWhiteSpace(overrideRoot))
        {
            return new DirectoryInfo(overrideRoot);
        }

        var candidates = new[]
        {
            new DirectoryInfo(Environment.CurrentDirectory),
            new DirectoryInfo(AppContext.BaseDirectory),
        };

        foreach (var start in candidates)
        {
            for (var current = start; current is not null; current = current.Parent)
            {
                if (File.Exists(Path.Combine(current.FullName, "pyproject.toml")))
                {
                    return current;
                }
            }
        }

        throw new BackendException("Could not find the project root containing pyproject.toml.");
    }

    private static string? FindRepoPython(DirectoryInfo root)
    {
        var venvPython = Path.Combine(root.FullName, ".venv", "Scripts", "python.exe");
        if (File.Exists(venvPython))
        {
            return venvPython;
        }

        return null;
    }

    private static string FindUvExecutable()
    {
        var explicitUv = Environment.GetEnvironmentVariable("UV");
        if (!string.IsNullOrWhiteSpace(explicitUv))
        {
            return explicitUv;
        }

        foreach (var path in (Environment.GetEnvironmentVariable("PATH") ?? "")
            .Split(Path.PathSeparator, StringSplitOptions.RemoveEmptyEntries))
        {
            var candidate = Path.Combine(path.Trim(), "uv.exe");
            if (File.Exists(candidate))
            {
                return candidate;
            }
        }

        var localAppData = Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData);
        var windowsAppsUv = Path.Combine(localAppData, "Microsoft", "WindowsApps", "uv.exe");
        if (File.Exists(windowsAppsUv))
        {
            return windowsAppsUv;
        }

        var wingetPackages = Path.Combine(localAppData, "Microsoft", "WinGet", "Packages");
        if (Directory.Exists(wingetPackages))
        {
            var packageUv = Directory
                .EnumerateFiles(wingetPackages, "uv.exe", SearchOption.AllDirectories)
                .FirstOrDefault(path => path.Contains("astral-sh.uv", StringComparison.OrdinalIgnoreCase));
            if (packageUv is not null)
            {
                return packageUv;
            }
        }

        return "uv";
    }

    private Guid Track(Process process)
    {
        var id = Guid.NewGuid();
        _processes[id] = process;
        return id;
    }

    private static void Terminate(Process process)
    {
        try
        {
            if (!process.HasExited)
            {
                process.Kill(entireProcessTree: true);
            }
        }
        catch (InvalidOperationException)
        {
        }
    }

    private static string? LastJsonLine(string output)
    {
        return output
            .Split(['\r', '\n'], StringSplitOptions.RemoveEmptyEntries)
            .LastOrDefault(line => line.TrimStart().StartsWith('{'));
    }

    private static void ThrowIfFailure(string json, string fallback)
    {
        var failure = JsonSerializer.Deserialize<FailureEnvelope>(json, JsonOptions);
        if (failure?.Event == "failed")
        {
            throw new BackendException(failure.Error ?? fallback);
        }
    }
}
